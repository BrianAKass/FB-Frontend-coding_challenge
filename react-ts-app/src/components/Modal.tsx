import React, { ReactElement, useState } from "react";
import { gql, useMutation } from "@apollo/client";
import "../styles/Modal.css";

const EDIT_PERSON = gql`
  mutation UpdatePerson($email: String!, $payload: EditPerson) {
    editPerson(email: $email, payload: $payload) {
      email
      name {
        title
        first
        last
      }
    }
  }
`;

interface Props {
  title?: string;
  first?: string;
  last?: string;

  email?: string;
  edit?: Function;
  handleOpen: Function;
}

export default function Modal({
  title,
  first,
  last,
  email,

  handleOpen,
}: Props): ReactElement {
  const initialState = {
    person_title: title,
    person_first: first,
    person_last: last,
    person_email: email,
  };

  const [updateData] = useMutation(EDIT_PERSON);
  const [newVal, setNewVal] = useState(initialState);
  const { person_title, person_first, person_last, person_email } = newVal;
  const handleChange = (field: string) => (e: any) => {
    setNewVal({ ...newVal, [field]: e.target.value });
  };

  return (
    <div className="modalContainer">
      <div className="modal">
        <h1>Edit User Data:</h1>
        <form
          onSubmit={(e) => {
            e.preventDefault();
            updateData({
              variables: {
                email: email,
                payload: {
                  email: person_email,
                  title: person_title,
                  first: person_first,
                  last: person_last,
                },
              },
              update: (cache, { data: updateData }) => {
                cache.modify({
                  id: cache.identify({
                    __typename: "Person",
                    email: email,
                  }),
                  fields: {
                    email: () => updateData.editPerson.email,
                    title: () => updateData.editPerson.title,
                    first: () => updateData.editPerson.first,
                    last: () => updateData.editPerson.last,
                  },
                });
              },
            });
            handleOpen();
          }}
        >
          <label>Title</label>
          <input
            className="modal-input"
            value={person_title}
            onChange={handleChange("person_title")}
          />
          <label>First Name</label>
          <input
            className="modal-input"
            value={person_first}
            onChange={handleChange("person_first")}
          />
          <label>Last Name</label>
          <input
            className="modal-input"
            value={person_last}
            onChange={handleChange("person_last")}
          />
          <label>Email</label>
          <input
            className="modal-input"
            value={person_email}
            onChange={handleChange("person_email")}
          />
          <div>
            <button
              className="modal-button modal-close"
              data-toggle="tooltip"
              data-placement="left"
              title="Close without changes"
              onClick={() => handleOpen()}
            >
              <i className="fas fa-times" /> Close
            </button>
            <button
              type="submit"
              className="modal-button modal-edit"
              data-toggle="tooltip"
              data-placement="right"
              title="Edit User Data"
            >
              <i className="fas fa-edit" /> Edit
            </button>
          </div>
        </form>
      </div>
    </div>
  );
}
