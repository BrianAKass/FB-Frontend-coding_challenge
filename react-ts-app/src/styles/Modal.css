.modal {
  position: fixed;
  z-index: 3;
  min-width: 230px;
  max-width: 315px;
  height: 300px;
  width: 100%;
  background: #f4f4f4;
  text-align: left;
  line-height: 1.4em;
  border-radius: 5%;
  box-shadow: 0 0 0 1px #e4e4e4;
  margin: 1rem;

  /* modal stuff */

  padding: 0 25px;
  position: absolute;
  left: calc(50% - 200px);
  top: 20%;
  border-radius: 4px;
  z-index: 999;
  pointer-events: auto;
  cursor: auto;
  box-shadow: 0 3px 7px rgba(0, 0, 0, 0.6);
}

.modalContainer {
  background-color: rgba(0, 0, 0, 0.6);
  width: 100vw;
  height: 100vh;
  position: fixed;
  left: 0;
  top: 0;
  z-index: 9;
  pointer-events: none;
  opacity: 1;
}

.modal-input {
  width: 100%;
  box-sizing: border-box;
}

.modal-button {
  padding: 5px;
  border: 1px solid #000;
  border-radius: 5px;
  font-size: 1em;
  text-transform: uppercase;
  margin: 10px 0;
  display: inline-block;
  opacity: 0.65;
  width: 47%;
  text-align: center;
  text-decoration: none;
  font-weight: 600;
  letter-spacing: 1px;
}

.modal-button:hover {
  opacity: 1;
}

.modal-close {
  background: #333;
  color: #fff;
  margin-right: 7px;
}

.modal-edit {
  background: #fff;
  color: #000;
  margin-left: 7px;
}
.hide {
  display: hidden;
}
