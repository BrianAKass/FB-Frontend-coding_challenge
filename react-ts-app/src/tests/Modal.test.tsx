import { render } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import Modal, { Props, EDIT_PERSON } from "../components/Modal";
import { testPeople } from "./Testpeople";

const defaultProps: Props = {
  title: testPeople[0].name.title,
  first: testPeople[0].name.first,
  last: testPeople[0].name.last,
  email: testPeople[0].email,
  handleOpen() {
    return;
  },
};
const mocks = [
  {
    request: {
      query: EDIT_PERSON,
      variables: {
        email: testPeople[0].email,
        payload: {
          email: testPeople[0].email,
          title: testPeople[0].name.title,
          first: testPeople[0].name.first,
          last: testPeople[0].name.last,
        },
      },
    },
  },
];
describe("<Modal/>", () => {
  it("Modal rendered AND EDIT_USER Mutation is working", () => {
    render(
      <MockedProvider mocks={mocks}>
        <Modal {...defaultProps} />
      </MockedProvider>
    );
  });
});
