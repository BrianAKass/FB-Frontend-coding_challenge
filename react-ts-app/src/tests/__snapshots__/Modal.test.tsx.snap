// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal/> Modal rendered AND EDIT_USER Mutation is working 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="modalContainer"
      >
        <div
          class="modal"
        >
          <h1>
            Edit User Data:
          </h1>
          <form>
            <label>
              Title
            </label>
            <input
              class="modal-input"
              value="Mr"
            />
            <label>
              First Name
            </label>
            <input
              class="modal-input"
              value="Emilio"
            />
            <label>
              Last Name
            </label>
            <input
              class="modal-input"
              value="Alvarez"
            />
            <label>
              Email
            </label>
            <input
              class="modal-input"
              value="emilio.alvarez@example.com"
            />
            <div>
              <button
                class="modal-button modal-close"
                data-placement="left"
                data-toggle="tooltip"
                title="Close without changes"
              >
                <i
                  class="fas fa-times"
                />
                 Close
              </button>
              <button
                class="modal-button modal-edit"
                data-placement="right"
                data-toggle="tooltip"
                title="Edit User Data"
                type="submit"
              >
                <i
                  class="fas fa-edit"
                />
                 Edit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="modalContainer"
    >
      <div
        class="modal"
      >
        <h1>
          Edit User Data:
        </h1>
        <form>
          <label>
            Title
          </label>
          <input
            class="modal-input"
            value="Mr"
          />
          <label>
            First Name
          </label>
          <input
            class="modal-input"
            value="Emilio"
          />
          <label>
            Last Name
          </label>
          <input
            class="modal-input"
            value="Alvarez"
          />
          <label>
            Email
          </label>
          <input
            class="modal-input"
            value="emilio.alvarez@example.com"
          />
          <div>
            <button
              class="modal-button modal-close"
              data-placement="left"
              data-toggle="tooltip"
              title="Close without changes"
            >
              <i
                class="fas fa-times"
              />
               Close
            </button>
            <button
              class="modal-button modal-edit"
              data-placement="right"
              data-toggle="tooltip"
              title="Edit User Data"
              type="submit"
            >
              <i
                class="fas fa-edit"
              />
               Edit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
